// ts-sql-query/queryRunners/ConsoleLogNoopQueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleLogNoopQueryRunner = void 0;
class ConsoleLogNoopQueryRunner {
    constructor(databaseOrConfig = 'noopDB') {
        if (typeof databaseOrConfig === 'string') {
            databaseOrConfig = { database: databaseOrConfig };
        }
        this.database = databaseOrConfig.database || 'noopDB';
        this.promise = databaseOrConfig.promise || Promise;
    }
    useDatabase(database) {
        // @ts-ignore
        this.database = database;
    }
    getNativeRunner() {
        return null;
    }
    execute(fn) {
        return fn(null);
    }
    executeSelectOneRow(query, params = []) {
        console.log('executeSelectOneRow:', query, params);
        return this.promise.resolve(undefined);
    }
    executeSelectManyRows(query, params = []) {
        console.log('executeSelectManyRows:', query, params);
        return this.promise.resolve([]);
    }
    executeSelectOneColumnOneRow(query, params = []) {
        console.log('executeSelectOneColumnOneRow:', query, params);
        return this.promise.resolve(undefined);
    }
    executeSelectOneColumnManyRows(query, params = []) {
        console.log('executeSelectOneColumnManyRows:', query, params);
        return this.promise.resolve([]);
    }
    executeInsert(query, params = []) {
        console.log('executeInsert:', query, params);
        return this.promise.resolve(0);
    }
    executeInsertReturningLastInsertedId(query, params = []) {
        console.log('executeInsertReturningLastInsertedId:', query, params);
        return this.promise.resolve(undefined);
    }
    executeInsertReturningMultipleLastInsertedId(query, params = []) {
        console.log('executeInsertReturningMultipleLastInsertedId:', query, params);
        return this.promise.resolve([]);
    }
    executeUpdate(query, params = []) {
        console.log('executeUpdate:', query, params);
        return this.promise.resolve(0);
    }
    executeDelete(query, params = []) {
        console.log('executeDelete:', query, params);
        return this.promise.resolve(0);
    }
    executeProcedure(query, params = []) {
        console.log('executeProcedure:', query, params);
        return this.promise.resolve();
    }
    executeFunction(query, params = []) {
        console.log('executeFunction:', query, params);
        return this.promise.resolve(undefined);
    }
    executeBeginTransaction() {
        console.log('executeBeginTransaction:', undefined, undefined);
        return this.promise.resolve();
    }
    executeCommit() {
        console.log('executeCommit:', undefined, undefined);
        return this.promise.resolve();
    }
    executeRollback() {
        console.log('executeRollback:', undefined, undefined);
        return this.promise.resolve();
    }
    executeDatabaseSchemaModification(query, params = []) {
        console.log('executeDatabaseSchemaModification:', query, params);
        return this.promise.resolve();
    }
    addParam(params, value) {
        const index = params.length;
        let result;
        switch (this.database) {
            case 'mariaDB':
                result = '?';
                break;
            case 'mySql':
                result = '?';
                break;
            case 'noopDB':
                result = '$' + index;
                break;
            case 'oracle':
                result = ':' + index;
                break;
            case 'postgreSql':
                result = '$' + (index + 1);
                break;
            case 'sqlite':
                result = '?';
                break;
            case 'sqlServer':
                result = '@' + index;
                break;
            default:
                throw new Error('Unknown database ' + this.database);
        }
        params.push(value);
        return result;
    }
    addOutParam(params, name) {
        const index = params.length;
        params.push({ out_param_with_name: name });
        return ':' + index;
    }
    createResolvedPromise(result) {
        return this.promise.resolve(result);
    }
}
exports.ConsoleLogNoopQueryRunner = ConsoleLogNoopQueryRunner;
