// ts-sql-query/queryRunners/MariaDBPoolQueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MariaDBPoolQueryRunner = void 0;
const AbstractPoolQueryRunner_1 = require("./AbstractPoolQueryRunner");
const MariaDBQueryRunner_1 = require("./MariaDBQueryRunner");
class MariaDBPoolQueryRunner extends AbstractPoolQueryRunner_1.AbstractPoolQueryRunner {
    constructor(pool, database = 'mariaDB') {
        super();
        this.pool = pool;
        this.database = database;
    }
    useDatabase(database) {
        if (database !== 'mariaDB' && database !== 'mySql') {
            throw new Error('Unsupported database: ' + database + '. MariaDBQueryRunner only supports mariaDB or mySql databases');
        }
        else {
            // @ts-ignore
            this.database = database;
        }
    }
    getNativeRunner() {
        return this.pool;
    }
    addParam(params, value) {
        params.push(value);
        return '?';
    }
    addOutParam(_params, _name) {
        throw new Error('Unsupported output parameters');
    }
    executeInsertReturningMultipleLastInsertedId(_query, _params = []) {
        throw new Error('Unsupported executeInsertReturningLastInsertedId for this database');
    }
    createResolvedPromise(result) {
        return Promise.resolve(result);
    }
    createQueryRunner() {
        return this.pool.getConnection().then(mariaDBConnection => new MariaDBQueryRunner_1.MariaDBQueryRunner(mariaDBConnection, this.database));
    }
    releaseQueryRunner(queryRunner) {
        queryRunner.getNativeRunner().release();
    }
}
exports.MariaDBPoolQueryRunner = MariaDBPoolQueryRunner;
