// ts-sql-query/queryRunners/MsNodeSqlV8QueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsNodeSqlV8QueryRunner = void 0;
class MsNodeSqlV8QueryRunner {
    constructor(connection) {
        this.connection = connection;
        this.database = 'sqlServer';
    }
    useDatabase(database) {
        if (database !== 'sqlServer') {
            throw new Error('Unsupported database: ' + database + '. MsNodeSqlV8QueryRunner only supports sqlServer databases');
        }
    }
    getNativeRunner() {
        return this.connection;
    }
    execute(fn) {
        return fn(this.connection);
    }
    executeSelectOneRow(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve(undefined);
                }
                else if (rows.length > 1) {
                    reject(new Error('Too many rows, expected only zero or one row'));
                }
                else {
                    resolve(rows[0]);
                }
            });
        });
    }
    executeSelectManyRows(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve([]);
                }
                else {
                    resolve(rows);
                }
            });
        });
    }
    executeSelectOneColumnOneRow(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve(undefined);
                }
                else if (rows.length > 1) {
                    reject(new Error('Too many rows, expected only zero or one row'));
                }
                else {
                    const row = rows[0];
                    if (row) {
                        const columns = Object.getOwnPropertyNames(row);
                        if (columns.length > 1) {
                            throw new Error('Too many columns, expected only one column');
                        }
                        resolve(row[columns[0]]); // Value in the row of the first column without care about the name
                        return;
                    }
                    resolve(undefined);
                }
            });
        });
    }
    executeSelectOneColumnManyRows(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve([]);
                }
                else {
                    const result = [];
                    for (let i = 0, length = rows.length; i < length; i++) {
                        const row = rows[i];
                        const columns = Object.getOwnPropertyNames(row);
                        if (columns.length > 1) {
                            reject(new Error('Too many columns, expected only one column'));
                            return;
                        }
                        result.push(row[columns[0]]); // Value in the row of the first column without care about the name
                    }
                    resolve(result);
                }
            });
        });
    }
    executeInsert(query, params = []) {
        return new Promise((resolve, reject) => {
            let rowCount = 0;
            this.connection.query(query, params, function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(rowCount);
                }
            }).on('rowcount', function (count) {
                rowCount = count;
            });
        });
    }
    executeInsertReturningLastInsertedId(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve(new Error('Unable to find the last inserted id'));
                }
                else if (rows.length > 1) {
                    reject(new Error('Too many rows, expected only zero or one row'));
                }
                else {
                    const row = rows[0];
                    if (row) {
                        const columns = Object.getOwnPropertyNames(row);
                        if (columns.length > 1) {
                            throw new Error('Too many columns, expected only one column');
                        }
                        resolve(row[columns[0]]); // Value in the row of the first column without care about the name
                        return;
                    }
                    resolve(new Error('Unable to find the last inserted id'));
                }
            });
        });
    }
    executeInsertReturningMultipleLastInsertedId(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve(new Error('Unable to find the last inserted id'));
                }
                else {
                    const result = rows.map((row) => {
                        const columns = Object.getOwnPropertyNames(row);
                        if (columns.length > 1) {
                            throw new Error('Too many columns, expected only one column');
                        }
                        return row[columns[0]]; // Value in the row of the first column without care about the name
                    });
                    resolve(result);
                }
            });
        });
    }
    executeUpdate(query, params = []) {
        return new Promise((resolve, reject) => {
            let rowCount = 0;
            this.connection.query(query, params, function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(rowCount);
                }
            }).on('rowcount', function (count) {
                rowCount = count;
            });
        });
    }
    executeDelete(query, params = []) {
        return new Promise((resolve, reject) => {
            let rowCount = 0;
            this.connection.query(query, params, function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(rowCount);
                }
            }).on('rowcount', function (count) {
                rowCount = count;
            });
        });
    }
    executeProcedure(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(undefined);
                }
            });
        });
    }
    executeFunction(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error, rows) {
                if (error) {
                    reject(error);
                }
                else if (!rows) {
                    resolve(undefined);
                }
                else if (rows.length > 1) {
                    reject(new Error('Too many rows, expected only zero or one row'));
                }
                else {
                    const row = rows[0];
                    if (row) {
                        const columns = Object.getOwnPropertyNames(row);
                        if (columns.length > 1) {
                            throw new Error('Too many columns, expected only one column');
                        }
                        resolve(row[columns[0]]); // Value in the row of the first column without care about the name
                        return;
                    }
                    resolve(undefined);
                }
            });
        });
    }
    executeBeginTransaction() {
        return new Promise((resolve, reject) => {
            this.connection.beginTransaction(function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve();
                }
            });
        });
    }
    executeCommit() {
        return new Promise((resolve, reject) => {
            this.connection.commit(function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve();
                }
            });
        });
    }
    executeRollback() {
        return new Promise((resolve, reject) => {
            this.connection.rollback(function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve();
                }
            });
        });
    }
    executeDatabaseSchemaModification(query, params = []) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, function (error) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(undefined);
                }
            });
        });
    }
    addParam(params, value) {
        const index = params.length;
        params.push(value);
        return '@' + index;
    }
    addOutParam(_params, _name) {
        throw new Error('Unsupported output parameters');
    }
    createResolvedPromise(result) {
        return Promise.resolve(result);
    }
}
exports.MsNodeSqlV8QueryRunner = MsNodeSqlV8QueryRunner;
