// ts-sql-query/queryRunners/MySql2PoolQueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySql2PoolQueryRunner = void 0;
const AbstractPoolQueryRunner_1 = require("./AbstractPoolQueryRunner");
const MySql2QueryRunner_1 = require("./MySql2QueryRunner");
class MySql2PoolQueryRunner extends AbstractPoolQueryRunner_1.AbstractPoolQueryRunner {
    constructor(pool, database = 'mySql') {
        super();
        this.pool = pool;
        this.database = database;
    }
    useDatabase(database) {
        if (database !== 'mariaDB' && database !== 'mySql') {
            throw new Error('Unsupported database: ' + database + '. MySql2PoolQueryRunner only supports mySql or mariaDB databases');
        }
        else {
            // @ts-ignore
            this.database = database;
        }
    }
    getNativeRunner() {
        return this.pool;
    }
    addParam(params, value) {
        params.push(value);
        return '?';
    }
    addOutParam(_params, _name) {
        throw new Error('Unsupported output parameters');
    }
    executeInsertReturningMultipleLastInsertedId(_query, _params = []) {
        throw new Error('Unsupported executeInsertReturningLastInsertedId for this database');
    }
    createResolvedPromise(result) {
        return Promise.resolve(result);
    }
    createQueryRunner() {
        return new Promise((resolve, reject) => {
            this.pool.getConnection((error, mysql2Connection) => {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(new MySql2QueryRunner_1.MySql2QueryRunner(mysql2Connection, this.database));
                }
            });
        });
    }
    releaseQueryRunner(queryRunner) {
        queryRunner.getNativeRunner().release();
    }
}
exports.MySql2PoolQueryRunner = MySql2PoolQueryRunner;
