// ts-sql-query/queryRunners/PgPoolQueryRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgPoolQueryRunner = void 0;
const AbstractPoolQueryRunner_1 = require("./AbstractPoolQueryRunner");
const PgQueryRunner_1 = require("./PgQueryRunner");
class PgPoolQueryRunner extends AbstractPoolQueryRunner_1.AbstractPoolQueryRunner {
    constructor(pool) {
        super();
        this.pool = pool;
        this.database = 'postgreSql';
    }
    useDatabase(database) {
        if (database !== 'postgreSql') {
            throw new Error('Unsupported database: ' + database + '. PgPoolQueryRunner only supports postgreSql databases');
        }
    }
    getNativeRunner() {
        return this.pool;
    }
    addParam(params, value) {
        params.push(value);
        return '$' + params.length;
    }
    addOutParam(_params, _name) {
        throw new Error('Unsupported output parameters');
    }
    createResolvedPromise(result) {
        return Promise.resolve(result);
    }
    createQueryRunner() {
        return this.pool.connect().then(connection => new PgQueryRunner_1.PgQueryRunner(connection));
    }
    releaseQueryRunner(queryRunner) {
        queryRunner.getNativeRunner().release();
    }
}
exports.PgPoolQueryRunner = PgPoolQueryRunner;
