// ts-sql-query/View.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.View = void 0;
const ColumnImpl_1 = require("./internal/ColumnImpl");
const symbols_1 = require("./utils/symbols");
class ViewOf {
    constructor(name) {
        // @ts-ignore
        this.__type = 'view';
        this.__name = name;
    }
    as(as) {
        const result = new (this.constructor)();
        result.__as = as;
        return result;
    }
    forUseInLeftJoin() {
        return this;
    }
    forUseInLeftJoinAs(as) {
        return this.as(as);
    }
    column(name, type, adapter, adapter2) {
        if (typeof adapter === 'string') {
            return new ColumnImpl_1.ColumnImpl(this, name, adapter, adapter2);
        }
        return new ColumnImpl_1.ColumnImpl(this, name, type, adapter);
    }
    optionalColumn(name, type, adapter, adapter2) {
        if (typeof adapter === 'string') {
            return (new ColumnImpl_1.ColumnImpl(this, name, adapter, adapter2)).__asOptionalColumn();
        }
        return (new ColumnImpl_1.ColumnImpl(this, name, type, adapter)).__asOptionalColumn();
    }
    // @ts-ignore
    __addWiths(_withs) {
        // Do nothing
    }
}
class View extends ViewOf {
    constructor(name) {
        super(name);
    }
}
exports.View = View;
