// ts-sql-query/expressions/values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asValueSource = exports.Argument = exports.__getValueSourcePrivate = exports.isValueSource = void 0;
const symbols_1 = require("../utils/symbols");
function isValueSource(value) {
    if (value === undefined || value === null) {
        return false;
    }
    if (typeof value === 'object') {
        return typeof value.__isResultOptional === 'function';
    }
    return false;
}
exports.isValueSource = isValueSource;
function __getValueSourcePrivate(valueSource) {
    return valueSource;
}
exports.__getValueSourcePrivate = __getValueSourcePrivate;
class Argument {
    constructor(argumentType, typeName, required, mode, adapter) {
        this.type = argumentType;
        this.typeName = typeName;
        this.required = required;
        this.mode = mode;
        this.adapter = adapter;
    }
}
exports.Argument = Argument;
function asValueSource(valueSource) {
    return valueSource;
}
exports.asValueSource = asValueSource;
