// ts-sql-query/queryBuilders/FragmentQueryBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FragmentFunctionBuilderIfValue = exports.FragmentFunctionBuilder = exports.FragmentQueryBuilder = void 0;
const ValueSourceImpl_1 = require("../internal/ValueSourceImpl");
class FragmentQueryBuilder {
    constructor(type, optional, adapter) {
        this.__type = type;
        this.__adapter = adapter;
        this.__isOptional = optional;
    }
    sql(sql, ...params) {
        return new ValueSourceImpl_1.FragmentValueSource(this.__isOptional, sql, params, this.__type, this.__adapter);
    }
}
exports.FragmentQueryBuilder = FragmentQueryBuilder;
class FragmentFunctionBuilder {
    constructor(definitions) {
        this.definitions = definitions;
    }
    as(impl) {
        return (...args) => {
            const newArgs = [];
            for (let i = 0, length = args.length; i < length; i++) {
                const arg = args[i];
                if (arg instanceof ValueSourceImpl_1.ValueSourceImpl) {
                    newArgs.push(arg);
                }
                else {
                    const definition = this.definitions[i];
                    const optional = definition.required === 'optional';
                    const newArg = new ValueSourceImpl_1.SqlOperationConstValueSource(optional, arg, definition.typeName, definition.adapter);
                    newArgs.push(newArg);
                }
            }
            return impl.apply(undefined, newArgs);
        };
    }
}
exports.FragmentFunctionBuilder = FragmentFunctionBuilder;
class FragmentFunctionBuilderIfValue {
    constructor(sqlBuilderSource, definitions) {
        this.sqlBuilderSource = sqlBuilderSource;
        this.definitions = definitions;
    }
    as(impl) {
        return (...args) => {
            const newArgs = [];
            for (let i = 0, length = args.length; i < length; i++) {
                const arg = args[i];
                if (arg instanceof ValueSourceImpl_1.ValueSourceImpl) {
                    newArgs.push(arg);
                }
                else {
                    const definition = this.definitions[i];
                    const optional = definition.required === 'optional';
                    const valueMode = definition.mode === 'value';
                    if (optional && valueMode) {
                        if (!this.sqlBuilderSource.__sqlBuilder._isValue(arg)) {
                            return new ValueSourceImpl_1.SqlOperationValueSourceIfValueAlwaysNoop();
                        }
                    }
                    const newArg = new ValueSourceImpl_1.SqlOperationConstValueSource(optional, arg, definition.typeName, definition.adapter);
                    newArgs.push(newArg);
                }
            }
            return impl.apply(undefined, newArgs);
        };
    }
}
exports.FragmentFunctionBuilderIfValue = FragmentFunctionBuilderIfValue;
